### PR TITLE
fix: instance[key] should not change after updating failed

### PR DESCRIPTION
**Pull request by JimmyDaddy**
fix: instance[key] should not change after updating failed
eg.
```js
const user = new User({
    email: 'a@e.com',
    nickname: 'sss',
    fingerprint: 'finger111'
});
try {
    // assume it will be failed
    await user.update({ fingerprint: 'aaa' });
} catch (err) {}

user.fingerprint // should be 'finger111', not 'aaa'


```

### CODE REVIEW

File: lib/bone.js [hunk 1]
```diff
@@ -160,10 +160,26 @@ class Bone {
    */
   _validateAttributes(values = {}) {
     const { attributes } = this.constructor;
-    Object.keys(this.raw).map((valueKey) => {
+    // check changed values
+    const changes = this.changes();
+    let changedValue = {};
+    for (const key in changes) {
+      if (changes[key].length && changes[key].length === 2) {
```
Comment (cyjake) [inline lib/bone.js:167]:
```suggestion
      if (changes[key].length === 2) {
```

`changes[key].length &&` seems unnecessary

File: lib/bone.js [hunk 2]
```diff
@@ -160,10 +160,26 @@ class Bone {
    */
   _validateAttributes(values = {}) {
     const { attributes } = this.constructor;
-    Object.keys(this.raw).map((valueKey) => {
+    // check changed values
+    const changes = this.changes();
+    let changedValue = {};
```
Comment (cyjake) [inline lib/bone.js:165]: changedValues

File: lib/bone.js [hunk 3]
```diff
@@ -160,10 +160,26 @@ class Bone {
...
+      if (changes[key].length && changes[key].length === 2) {
+        changedValue[key] = changes[key][1];
+      }
+    }
+
+    // merge all changed values
+    changedValue = {
+      ...changedValue,
+      ...values,
+    };
```
Comment (cyjake) [inline lib/bone.js:176]: Personally I'd prefer `Object.assign(changedValues, values)`

File: lib/bone.js [hunk 4]
```diff
@@ -160,10 +160,26 @@ class Bone {
...
+
+    Object.keys(changedValue).map((valueKey) => {
       const attribute = attributes[valueKey];
+      if (!attribute) return;
```
Comment (cyjake) [inline lib/bone.js:180]:
```suggestion
    for (const valueKey of Object.keys(changedValue)) {
      const attribute = attributes[valueKey];
      if (attribute == null) return;
```

File: lib/validator.js [hunk 5]
```diff
@@ -49,6 +49,9 @@ const validators = {
   is(str, pattern, modifiers) {
     return this.regex(str, pattern, modifiers);
   },
+  notEmpty(str) {
+    return !str.match(/^[\s\t\r\n]*$/);
```
Comment (cyjake) [inline lib/validator.js:53]: no need to match the complete string, test for non whitespace character shall suffice

```suggestion
    return /[^\s]/.test(str);
```

Comment (cyjake) [reply lib/validator.js:53]: btw, `\s` includes `[\t\r\n]`

File: test/integration/suite/basics.test.js [hunk 6]
```diff
@@ -306,6 +307,13 @@ describe('=> Accessors', function() {
     expect(user.status).to.eql(1);
     expect(user.raw.status).to.equal(-1);
     await user.update({ status: 2 });
+    //set status(value = 0) {
```
Comment (cyjake) [inline test/integration/suite/basics.test.js:310]:
```suggestion
    // set status(value = 0) {
```

**Review by cyjake — APPROVED** [review]
+1
